lib/auth/oidc_google.go

```diff
@@ -0,0 +1,192 @@
+
+		uri, err := utils.ParseSessionsURI(connector.GetGoogleServiceAccountURI())
+		if err != nil {
+			return nil, trace.BadParameter("failed to parse google_service_account_uri: %v", err)
+		}
+		jsonCredentials, err = os.ReadFile(uri.Path)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	} else if connector.GetGoogleServiceAccount() != "" {
+		// load the google service account from string
+		credentialLoadingMethod = "google_service_account"
+		jsonCredentials = []byte(connector.GetGoogleServiceAccount())
+	}
+
+	// The "Admin SDK Directory API" needs admin delegation (see
+	// https://developers.google.com/admin-sdk/directory/v1/guides/delegation
+	// and
+	// https://developers.google.com/identity/protocols/oauth2/service-account#delegatingauthority )
+	// and the "Cloud Identity API" needs an account with View permission on
+	// all groups to work reliably.
+	credentialsParams := google.CredentialsParams{
+		Subject: connector.GetGoogleAdminEmail(),
+	}
+
+	if connector.GetGoogleTransitiveGroups() {
+		log.Debugf("Loading credentials to fetch transitive groups")
+		credentialsParams.Scopes = []string{cloudidentity.CloudIdentityGroupsReadonlyScope}
+	} else {
+		log.Debugf("Loading credentials to fetch direct groups")
+		credentialsParams.Scopes = []string{directory.AdminDirectoryGroupReadonlyScope}
+	}
+
+	credentials, err := google.CredentialsFromJSONWithParams(ctx, jsonCredentials, credentialsParams)
+	if err != nil {
+		return nil, trace.BadParameter("unable to parse google service account from %v: %v", credentialLoadingMethod, err)
+	}
+
+	return credentials, nil
+}
+
+func groupsFromGsuiteDirectory(ctx context.Context, email string, clientOptions ...option.ClientOption) ([]string, error) {
+	service, err := directory.NewService(ctx, clientOptions...)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var groups []string
+	err = service.Groups.List().
+		UserKey(email).
+		Pages(ctx, func(resp *directory.Groups) error {
+			if resp == nil {
+				return nil
+			}
+			for _, g := range resp.Groups {
+				if g != nil && g.Email != "" {
+					groups = append(groups, g.Email)
+				}
+			}
+			return nil
+		})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return groups, nil
+}
+
+func groupsFromGsuiteCloudidentity(ctx context.Context, email string, clientOptions ...option.ClientOption) ([]string, error) {
+	service, err := cloudidentity.NewService(ctx, clientOptions...)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var groups []string
+	err = service.Groups.Memberships.SearchTransitiveGroups("groups/-").
+		// the google API docs claim that the query string is a CEL expression
+		// (https://opensource.google/projects/cel) but the call will fail if
+		// you use double quotes instead of single quotes in spite of them being
+		// equivalent according to the CEL specs
+		Query(fmt.Sprintf("member_key_id == '%s' && 'cloudidentity.googleapis.com/groups.discussion_forum' in labels", email)).
```

Comment:
Where did you get the values for the `SearchTransitiveGroups` and `Query`? Could you add a comment for those of us that aren't as familiar with the cloud identity api as you have become?